options/configurable.cc

```diff
@@ -0,0 +1,972 @@
+    bool input_strings_escaped, bool ignore_unused_options,
+    std::unordered_set<std::string>* unused_opts) {
+  Status s, result, invalid;
+  bool found_one = false;
+  std::unordered_set<std::string> invalid_opts;
+  std::string curr_opts;
+  if (!ignore_unused_options) {
+    // If we are not ignoring unused, get the defaults in case we need to reset
+    GetOptionString(&curr_opts, ";");
+  }
+  // Go through all of the values in the input map and attempt to configure the
+  // property.
+  for (const auto& o : opts_map) {
+    s = ConfigureOption(db_opts, o.first, o.second, input_strings_escaped);
+    if (s.ok()) {
+      found_one = true;
+    } else if (s.IsNotFound()) {
+      result = s;
+      unused_opts->insert(o.first);
+    } else if (s.IsNotSupported()) {
+      invalid_opts.insert(o.first);
+    } else {
+      invalid_opts.insert(o.first);
+      invalid = s;
+    }
+  }
+  // While there are unused properties and we processed at least one,
+  // go through the remaining unused properties and attempt to configure them.
+  while (found_one && !unused_opts->empty()) {
+    result = Status::OK();
+    found_one = false;
+    for (auto it = unused_opts->begin(); it != unused_opts->end();) {
+      s = ConfigureOption(db_opts, *it, opts_map.at(*it),
+                          input_strings_escaped);
+      if (s.ok()) {
+        found_one = true;
+        it = unused_opts->erase(it);
+      } else if (s.IsNotFound()) {
+        result = s;
+        ++it;
+      } else if (s.IsNotSupported()) {
+        invalid_opts.insert(*it);
+        it = unused_opts->erase(it);
+      } else {
+        invalid_opts.insert(*it);
+        it = unused_opts->erase(it);
+        invalid = s;
+      }
+    }
+  }
+  if (!invalid_opts.empty()) {
+    unused_opts->insert(invalid_opts.begin(), invalid_opts.end());
+  }
+  if (ignore_unused_options || (invalid.ok() && result.ok())) {
+    return Status::OK();
+  } else {
+    if (!curr_opts.empty()) {
+      // There are some options to reset from this current error
+      ConfigureFromString(db_opts, curr_opts, input_strings_escaped, true);
+    }
+    if (!invalid.ok()) {
+      return invalid;
+    } else {
+      return result;
+    }
+  }
+}
+
+Status Configurable::ConfigureFromMap(
+    const DBOptions& db_opts,
+    const std::unordered_map<std::string, std::string>& opts_map,
+    bool input_strings_escaped, bool ignore_unused_options) {
+  std::unordered_set<std::string> unused_opts;
+  Status s = DoConfigureFromMap(db_opts, opts_map, input_strings_escaped,
+                                ignore_unused_options, &unused_opts);
+  return s;
+}
+
+Status Configurable::ConfigureFromMap(
+    const DBOptions& db_opts,
+    const std::unordered_map<std::string, std::string>& opts_map,
+    bool input_strings_escaped, std::unordered_set<std::string>* unused) {
+  return DoConfigureFromMap(db_opts, opts_map, input_strings_escaped, true,
```

Comment:
Nit: we can add a piece of comment near constants to indicate what they mean. For example `/*ignore_unused_options=*/ true`.